docs/dev/relay_bring_your_own_codegen.rst

```diff
@@ -0,0 +1,529 @@
+..  Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+..    http://www.apache.org/licenses/LICENSE-2.0
+
+..  Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+=============================
+Bring Your Own Codegen To TVM
+=============================
+**Author**: `Zhi Chen <https://github.com/zhiics>`_, `Cody Hao Yu <https:://github.com/comaniac>`_
+
+As the number of hardware devices targeted by deep learning workloads keeps increasing, the required knowledge for users to achieve high performance on various devices keeps increasing as well. To free data scientists from worrying about the performance when developing a new model, hardware vendors either provide libraries such as MKLDNN or cuDNN with many commonly used deep learning operators, or provide frameworks such as TensorRT to let users describe their models in a certain way to achieve high performance. However, users have to learn a new programming interface when they attempt to work on a new library or device. As a result, the demand for a unified programming interface becomes more and more important to 1) let all users and hardware vendors stand on the same page, and 2) provide a feasible solution to allow specialized hardware or library to only support widely used operators with extremely high performance, but fallback unsupported operators to general devices like CPU/GPU.
+
+In this developer guide, we demonstrate how you, as a hardware vendor, can easily implement your own codegen and register it as a Relay backend compiler to support your hardware device/library. This guide covers two types of codegen based on different graph representations you need:
+
+**1. You want to generate C code.**
+
+If your hardware already has a well-optimized C/C++ library, such as Intel CBLAS/MKL to CPU and NVIDIA CUBLAS to GPU, then this is what you are looking for. Fortunately, C source code module is fully compatible with TVM runtime module, which means the generated code could be compiled by any C/C++ compiler with proper compilation flags, so the only task you have is to implement a codegen that generates C code for subgraphs and a C source module to integrate into TVM runtime module. We will demonstrate how to implement a C code generator for your hardware in the following section.
```

Comment:
consider update to talk about TVM_DLL_EXPORT_PACKED and TVM_DLL_EXPORT_TYPED

Reply:
I've tried but it seems cannot reduce the code size much. Since TVM_DLL_EXPORT_PACKED_FUNC packs args, type_code, num_args to TVMArgs, we still have to have another wrapper to unpack them back to TVMValue. As a result, [this wrapper](https://github.com/apache/incubator-tvm/blob/master/tests/python/relay/test_external_runtime.py#L147) is still required.

On the other hand, TVM_DLL_EXPORT_TYPED_FUNC does unpack TVMArgs according to native function signiture, but it seems can only return scalar values. Please correct me if I misused this marco.

Reply:
The API will looks like:
```c++
void json_rt_(DLTensor* x, DLTensor *y, DLTensor* out) {
   // followup code
} 

TVM_DLL_EXPORT_TYPED(json_rt, json_rt_); 
```

Reply:
Thanks for the advice. We can modify the C codegen to generate the following working code:

```c++
    extern "C" int json_rt_1_(DLTensor* arg0, DLTensor* arg1, DLTensor* arg2, DLTensor* arg3,
                              DLTensor* out) {
        // User provided function.
        gcc_1_(static_cast<float*>(arg0->data), static_cast<float*>(arg1->data),
                static_cast<float*>(arg2->data), static_cast<float*>(arg3->data),
                static_cast<float*>(out->data));
        return 0;
    }

    TVM_DLL_EXPORT_TYPED_FUNC(json_rt_1, json_rt_1_);
```

On the other hand, here is still a concern of including this marco. Since the macro is defined in `packed_func.h`, we need to include this file in the generated C code. However, `packed_func.h` includes `dmlc/logging.h`, which is not one of the default 3rd party include paths. As a result, users have to pass an additional argument for compilation:

```python
options=[ "-O2", "-std=c++11", "-I",  "./3rdparty/dmlc-core/include/"]
```

This seems not intuitive to users as they never use dmlc directly.

The solutions are: 1) Keep the current version. Since this wrapper is generated by ours, users will not see those complicated wrappers. 2) Put dmlc include path to libinfo, but it will increase the code size for all compiled modules. 3) Modify `packed_func.h` to make it dmlc indpendent. Please let us know which solution you prefer.

Reply:
I do think in this case perhaps it is better to introduce dep to dmlc-core and DLPack

Reply:
The specific macro handles quite a few cases that might be useful in c codegen packages (e.g. handle Object) 

Given that this is part of the TVM runtime, i feel it is OK to introduce the additional deps as long as they are deps of the tvm runtime

Reply:
OK we will first file a PR to include dmlc path and then come back to modify the codegen accordingly.

Reply:
see #4710

Reply:
changed accordingly. Please take another pass.